[Ingest Pipelines] Add route for structure tree

ElenaStoeva opened the pull request:
Addresses https://github.com/elastic/kibana/issues/227405

## Summary

This PR adds the `GET _ingest_pipelines/structure_tree/{root_pipeline_id}` route that returns Ingest Pipeline Structure tree of type `PipelineTreeNode` up to the 6th level.

**How to test?**
There is nothing to test locally as this route is not used in the plugin yet. The PR adds a good amount of unit and API integration tests that verify that the logic is correct.

**Why don't we use the existing list route (`GET _ingest_pipelines`) and transform the returned list to a `PipelineTreeNode` client-side?**
Elasticsearch returns the pipelines in a map structure (`{ [key: string]: estypes.IngestPipeline; }`) and the list route logic converts this to a list of pipelines (`Pipeline[]`). In the implementation of `fetchPipelineStructureTree`, we can fetch pipelines many times (sometimes the same pipeline can be fetched multiple times) and it is more efficient to fetch them from a map rather than from a list. Therefore, I thought it makes more sense to create a new route so that we can utilize the map structure that Es returns.

**Why do we fetch up to the 6th level and not up to the 5th level, given that the Structure Tree component displays up to 5 levels?**
The 5th level tree nodes in the component have a child node "+X more pipelines" so we need to know how many children the 5th level nodes have. Therefore we fetch the 6th level as well.

ElenaStoeva committed:
[Ingest Pipelines] Add route for structure tree

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/pipeline-structure-tree-routes

ElenaStoeva committed:
Add unit tests

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/pipeline-structure-tree-routes

ElenaStoeva committed:
Refactor logic and add api integration tests

ElenaStoeva committed:
Fix unit tests and type errors

ElenaStoeva committed:
Add docs

elasticmachine commented:
Pinging @elastic/kibana-management (Team:Kibana Management)

ElenaStoeva committed:
One more comment

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/320100)
* Commit: 6922cbbcdec5f4cf70c957031b5f80bb035980f0

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/8d764e977c4af52a6c25a3d042868231561ebae1) | [after](https://github.com/elastic/kibana/commit/6922cbbcdec5f4cf70c957031b5f80bb035980f0) | diff |
| --- | --- | --- | --- |
| `@kbn/ingest-pipelines-shared` | 1 | 7 | +6 |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/8d764e977c4af52a6c25a3d042868231561ebae1) | [after](https://github.com/elastic/kibana/commit/6922cbbcdec5f4cf70c957031b5f80bb035980f0) | diff |
| --- | --- | --- | --- |
| `@kbn/ingest-pipelines-shared` | 2 | 8 | +6 |


</details>


cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320100","number":320100,"commit":"6922cbbcdec5f4cf70c957031b5f80bb035980f0"}],"number":320100}
buildkite-pr-comment-->

sabarasaba reviewed (commented):
Nice job @ElenaStoeva! Left a few comments, mostly worried about having to fetch all pipelines

ElenaStoeva commented on x-pack/platform/plugins/shared/ingest_pipelines/server/routes/api/get.ts:
```diff
@@ -82,4 +83,36 @@ export const registerGetRoutes = ({ router, lib: { handleEsError } }: RouteDepen
+      const { name } = req.params;
+
+      try {
+        const allPipelines = await clusterClient.asCurrentUser.ingest.getPipeline();
+        const pipelineStructureTree = fetchPipelineStructureTree(allPipelines, name);
```
Yeah, I was thinking this too so that's why I added an API integration stress test to check if there is any issue. In the api integration test, I made a complex tree with 10 levels, and each node has 10 children. This means that on the 10th level, there are 10^10 nodes. The total number of unique pipelines is 100 (since each node on the same level has the same children as the other nodes on that level). The test passes and takes milliseconds. Do you think there might be a case where a user has more pipelines than that? I can change the implementation of the complex tree and make each node to have unique children, this will significantly increase the number of unique pipelines - it would be a geometric series 1 + 10 + 10^2 + ... + 10^N where N is the number of levels.

ElenaStoeva replied:
Updated the test tree in [2b4ade6](https://github.com/elastic/kibana/pull/228053/commits/2b4ade6b6ad72bfcfa701c6b05573f1859f8e856)

ElenaStoeva commented on x-pack/platform/plugins/shared/ingest_pipelines/server/routes/api/fetch_pipeline_structure_tree/fetch_pipeline_structure_tree.ts:
```diff
@@ -0,0 +1,54 @@
+ * @param rootPipelineId The pipeline
+ * @param level The level of the current {@link rootPipelineName} pipeline
+ */
+export const fetchPipelineStructureTree = (
```
I see your point... so we can have this `Map<string, PipelineTreeNode>` structure, which we pass as an argument, and in the beginning of the function we can check if a node already exists in this map, and if not, then process it. Is this what you meant?

sabarasaba replied:
Yeah that's what I had in mind!

ElenaStoeva committed:
Address feedback and refactor

ElenaStoeva commented:
Thank you very much for the review @sabarasaba! Here are the changes that I made in my last commit:

- Updated the logic of `fetchPipelineStructureTree` to use a map with created nodes. Also handled the case where the pipeline doesn't exist and added some more refactoring and simplification. Updated the unit tests as well.
- Updated the API integration test - made the test tree more complex with each node having unique children. Commited the tree with 7 levels and 3 children per node so that the test doesn't run for too long (and this is also a realistic case), but I ran it locally with 4 children per node and it still worked.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/320813)
* Commit: 2b4ade6b6ad72bfcfa701c6b05573f1859f8e856

### Failed CI Steps
* [Check Types](https://buildkite.com/elastic/kibana-pull-request/builds/320813#01981d61-118f-4e18-9278-547e9ddaace6)

### History
* :green_heart: [Build #320100](https://buildkite.com/elastic/kibana-pull-request/builds/320100) succeeded 6922cbbcdec5f4cf70c957031b5f80bb035980f0

cc @ElenaStoeva

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320813","number":320813,"commit":"2b4ade6b6ad72bfcfa701c6b05573f1859f8e856"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/320100","number":320100,"commit":"6922cbbcdec5f4cf70c957031b5f80bb035980f0"}],"number":320813}
buildkite-pr-comment-->

ElenaStoeva committed:
Fix type errors

ElenaStoeva committed:
Merge branch 'main' into ingest-pipelines/pipeline-structure-tree-routes

sabarasaba reviewed (approved):
latest lgtm
